src/promptflow/promptflow/azure/_storage/cosmosdb/summary.py

```diff
@@ -143,54 +171,71 @@ def _persist_line_run(self, client):
+    def _insert_evaluation_with_retry(self, client: ContainerProxy):
+        for attempt in range(3):
+            try:
+                # We receive requests to persist main flow first and then evaluation,
+                # but init cosmosDB client is time consuming, and the later request will reuse the same client,
+                # so it's possible that the main run is not persisted when we start to patch evaluation to it.
+                self._insert_evaluation(client)
+                break
+            except InsertEvaluationsRetriableException as e:
+                if attempt == 2:  # If this was the last attempt, ignore and just return
+                    current_app.logger.error(f"Error while inserting evaluation: {e}")
+                    return
+                time.sleep(1)
+
+    def _insert_evaluation(self, client: ContainerProxy):
         attributes: dict = self.span._content[SpanFieldName.ATTRIBUTES]
         partition_key = self.span.session_id
         name = self.span.name
         item = LineEvaluation(
             trace_id=self.span.trace_id,
             root_span_id=self.span.span_id,
-            outputs=json_loads_parse_const_as_str(attributes[SpanAttributeFieldName.OUTPUT]),
+            outputs=json_loads_parse_const_as_str(attributes.get(SpanAttributeFieldName.OUTPUT, "{}")),
             display_name=name,
             created_by=self.created_by,
         )
-        if SpanAttributeFieldName.REFERENCED_LINE_RUN_ID in attributes:
-            # Query to get item id from line run id.
-            line_run_id = attributes[SpanAttributeFieldName.REFERENCED_LINE_RUN_ID]
-            query = "SELECT * FROM c WHERE c.line_run_id = @line_run_id"
-            parameters = [
-                {"name": "@line_run_id", "value": line_run_id},
-            ]
-            items = list(client.query_items(query=query, parameters=parameters, partition_key=partition_key))
-            if items:
-                main_id = items[0]["id"]
-            else:
-                current_app.logger.error(f"Cannot find main run for line run id: {line_run_id}")
-                return
-            line_run_id = attributes[SpanAttributeFieldName.LINE_RUN_ID]
-            item.line_run_id = line_run_id
+
+        # None is the default value for the field.
+        referenced_line_run_id = attributes.get(SpanAttributeFieldName.REFERENCED_LINE_RUN_ID, None)
+        referenced_batch_run_id = attributes.get(SpanAttributeFieldName.REFERENCED_BATCH_RUN_ID, None)
+        line_number = attributes.get(SpanAttributeFieldName.LINE_NUMBER, None)
+
+        query = (
+            "SELECT * FROM c WHERE "
+            "c.line_run_id = @line_run_id AND c.batch_run_id = @batch_run_id AND c.line_number = @line_number"
+        )
+        parameters = [
+            {"name": "@line_run_id", "value": referenced_line_run_id},
+            {"name": "@batch_run_id", "value": referenced_batch_run_id},
+            {"name": "@line_number", "value": line_number},
```

Comment:
Do we need to check null for parameter before query?

Reply:
It's by design. Since we also save `batch_run_id` as "" to container.
I've tested we could get batch run/flow test main flow by this query.